File: number_theory/sieve_eratosthenes.cpp
// sieve_eratosthenes.cpp - Generate prime numbers up to N using the Sieve of Eratosthenes.
// Author: Scott

#include <iostream>
#include <vector>
using namespace std;

void sieve(int n) {
    vector<bool> is_prime(n+1, true);
    is_prime[0] = is_prime[1] = false;
    for (int p = 2; p*p <= n; ++p)
        if (is_prime[p])
            for (int i = p*p; i <= n; i += p)
                is_prime[i] = false;

    cout << "Primes up to " << n << ": ";
    for (int i = 2; i <= n; ++i)
        if (is_prime[i]) cout << i << " ";
    cout << endl;
}

int main() {
    int n = 100;
    cout << "Generate primes up to (default 100): ";
    string input;
    getline(cin, input);
    if (!input.empty()) n = stoi(input);

    sieve(n);
    return 0;
}

File: number_theory/sieve_eratosthenes.py
"""
sieve_eratosthenes.py - Generate all prime numbers up to a given limit using the Sieve of Eratosthenes.
Author: Scott
"""

def sieve(n):
    prime = [True] * (n+1)
    prime[0:2] = [False, False]
    for i in range(2, int(n**0.5)+1):
        if prime[i]:
            for j in range(i*i, n+1, i):
                prime[j] = False
    return [i for i, is_prime in enumerate(prime) if is_prime]

if __name__ == "__main__":
    n = int(input("Generate primes up to: ") or "100")
    primes = sieve(n)
    print(f"Primes up to {n}: {primes}")
